src/Physics/NuclearState/NucleusGenINCL.cxx

```diff
@@ -0,0 +1,336 @@
+
+
+  if(is_coh||is_ve) {
+    // ** For COH or ve- set a vertex positon on the nuclear boundary
+    //
+  //  LOG("Vtx", pINFO)  << "Setting vertex on the nuclear boundary";
+  //  double phi      = 2*kPi * rnd->RndFsi().Rndm();
+  //  double cosphi   = TMath::Cos(phi);
+  //  double sinphi   = TMath::Sin(phi);
+  //  double costheta = -1 + 2 * rnd->RndFsi().Rndm();
+  //  double sintheta = TMath::Sqrt(1-costheta*costheta);
+  //  vtx.SetX(R*sintheta*cosphi);
+  //  vtx.SetY(R*sintheta*sinphi);
+  //  vtx.SetZ(R*costheta);
+  }
+  else {
+    vtx = incl_nucleus->getHitNucleonPosition();
+  }
+
+  }
+
+  LOG("NucleusGenINCL", pINFO) << "Position";
+  vtx.Print();
+  // Copy the vertex info to the particles already in the event  record
+  //
+  TObjArrayIter piter(evrec);
+  GHepParticle * p = 0;
+  while( (p = (GHepParticle *) piter.Next()) )
+  {
+    if(pdg::IsPseudoParticle(p->Pdg())) continue;
+    if(pdg::IsIon           (p->Pdg())) continue;
+
+    LOG("NucleusGenINCL", pINFO) << "Setting vertex position for: " << p->Name();
+    p->SetPosition(vtx.x(), vtx.y(), vtx.z(), 0.);
+  }
+
+
+
+}
+
+void NucleusGenINCL::setInitialStateMomentum(GHepRecord * evrec) const{
+  Interaction *  interaction = evrec       -> Summary();
+  InitialState * init_state  = interaction -> InitStatePtr();
+  Target *       tgt         = init_state  -> TgtPtr();
+
+  // do nothing for non-nuclear targets
+  if(!tgt->IsNucleus()) return;
+
+  TLorentzVector * p4 = tgt->HitNucP4Ptr();
+
+  // do nothing if the struct nucleon 4-momentum was set (eg as part of the
+  // initial state selection)
+  if(p4->Px()>0 || p4->Py()>0 || p4->Pz()>0) return;
+
+  // access the hit nucleon and target nucleus at the GHEP record
+  GHepParticle * nucleon = evrec->HitNucleon();
+  GHepParticle * nucleus = evrec->TargetNucleus();
+  assert(nucleon);
+  assert(nucleus);
+
+  // initialize INCL nucleus model
+  // INCL nucleus model sample all nucleons with r-p correlation
+  INCLNucleus *incl_nucleus = INCLNucleus::Instance();
+  G4INCL::Nucleus *incl_nuc = incl_nucleus->getNuclues();
+  TLorentzVector p4tgt;
+  p4tgt.SetPx(incl_nuc->getMomentum().getX() / 1000.);
+  p4tgt.SetPy(incl_nuc->getMomentum().getY() / 1000. );
+  p4tgt.SetPz(incl_nuc->getMomentum().getZ() / 1000. );
+  p4tgt.SetE(incl_nuc->getEnergy() / 1000.);
+  init_state->SetTgtP4(p4tgt);
+  nucleus->SetMomentum(p4tgt);
+
+
+  // get a random nucleon with respect to the isospin of evrec->HitNucleon();
+  // the removal energy maybe not necessary
+  TVector3 p3 = incl_nucleus->getHitNucleonMomentum();
+  double   hit_nucleon_energy = incl_nucleus->getHitNucleonEnergy();
+  double   w  = incl_nucleus->getRemovalEnergy();
+  //-- update the struck nucleon 4p at the interaction summary and at
+  // the GHEP record
+  p4->SetPx(p3.Px()/1000.);
+  p4->SetPy(p3.Py()/1000.);
+  p4->SetPz(p3.Pz()/1000.);
+  p4->SetE (hit_nucleon_energy/1000.);
+  LOG("NucleusGenINCL", pINFO) << "Momentum";
+  p4->Print();
```

Comment:
Again consider `PrintUtils` `Vec3AsString()` so that can be part of the LOG line and controlled by the message service.